crates/sqlbuiltins/src/functions/scalars/kdl.rs

```diff
@@ -114,49 +126,7 @@ impl BuiltinScalarUDF for KDLMatches {
     }
 }
 
-fn kdl_parse_udf_args(
-    args: &[ColumnarValue],
-) -> datafusion::error::Result<(KdlDocument, KdlQuery)> {
-    // parse the filter first, because it's probably shorter and
-    // erroring earlier would be preferable to parsing a large that we
-    // don't need/want.
-    let filter: kdl::KdlQuery = match get_nth_scalar_value(args, 1) {
-        Some(ScalarValue::Utf8(Some(val))) | Some(ScalarValue::LargeUtf8(Some(val))) => {
-            val.parse().map_err(|err: ::kdl::KdlError| {
-                datafusion::common::DataFusionError::Execution(err.to_string())
-            })?
-        }
-        Some(val) => {
-            return Err(datafusion::common::DataFusionError::Execution(format!(
-                "invalid type for KQL expression {}",
-                val.data_type(),
-            )))
-        }
-        None => {
-            return Err(datafusion::common::DataFusionError::Execution(
-                "missing KQL query".to_string(),
-            ))
-        }
-    };
-
-    let doc: kdl::KdlDocument = match get_nth_scalar_value(args, 0) {
-        Some(ScalarValue::Utf8(Some(val))) | Some(ScalarValue::LargeUtf8(Some(val))) => {
-            val.parse().map_err(|err: ::kdl::KdlError| {
-                datafusion::common::DataFusionError::Execution(err.to_string())
-            })?
-        }
-        Some(val) => {
-            return Err(datafusion::common::DataFusionError::Execution(format!(
-                "invalid type for KDL value {}",
-                val.data_type(),
-            )))
-        }
-        None => {
-            return Err(datafusion::common::DataFusionError::Execution(
-                "invalid field for KDL".to_string(),
-            ))
-        }
-    };
-
-    Ok((doc, filter))
+#[memoize(Capacity: 256)]
+fn compile_kdl_query(query: String) -> Result<KdlQuery, BuiltinError> {
```

Comment:
This feels like a pre-optimization to me. 

we are already only compiling once per query, not once per element (row). If it were compiling per row, i could see the need for memoizing.

Reply:
no, we are compiling it once per row https://github.com/GlareDB/glaredb/pull/2304/files/29c1f5bcdb81461f20252262873bf5d9f8995a81#diff-01dbd7d78f91996d86c9808046c671a61277870e54a01ef26d996479b24aec90R45 <- this happens within a closure that gets called for every row of the column.